File: java/429.n叉树的层序遍历.java
/*
 * @lc app=leetcode.cn id=429 lang=java
 *
 * [429] N叉树的层序遍历
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
    private List<List<Integer>> res = new LinkedList<>() ;
    /**
     * 递归算法
     * 1ms   100%   82%   39.5MB
     */
    public List<List<Integer>> levelOrder1(Node root) {
        bfs(root,0);
        return res;
    }
    private void bfs(Node node,int level){
        if(node == null)return;
        if(res.size()==level){
            res.add(new LinkedList<Integer>());
        }
        res.get(level).add(node.val);
        for(Node n : node.children)bfs(n,level+1);
    }

    /**
     * 迭代算法，
     * 4ms   54%   76%   39.5MB
     */
    public List<List<Integer>> levelOrder(Node root) {
        if(root ==null) return  res;
        Deque<Node> queue = new LinkedList<>();
        queue.add(root);
        while(!queue.isEmpty()){
            int rowSize = queue.size();
            List<Integer> levelRes = new ArrayList<>(rowSize);
            for(int i=0;i<rowSize;i++){
                Node node = queue.poll();
                levelRes.add(node.val);
                for(Node n:node.children){
                    queue.offer(n);
                }
            }
            res.add(levelRes);
        }
        return res;
    }
}
// @lc code=end

File: py/429.n叉树的层序遍历.py
#
# @lc app=leetcode.cn id=429 lang=python3
#
# [429] N叉树的层序遍历
#

# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    # 递归算法
    # 64MS  67%   34.6%   15.3MB
    def levelOrder1(self, root: 'Node') -> List[List[int]]:
        res = []
        def bfs(node,level):
            if node is None:
                return
            if(len(res)==level):
                res.append([])
            res[level].append(node.val)
            for n in node.children :
                bfs(n, level+1)
        bfs(root,0)
        return res

# 迭代 队列
# 64ms   67%   32%  15.3MB
    def levelOrder(self, root: 'Node') -> List[List[int]]:
        res = []
        if root is None:
            return res
        queue = [root]
        while len(queue) !=0:
            row_size = len(queue)
            row_res = []
            for i in range(row_size):
                node = queue.pop(0)
                row_res.append(node.val)
                for n in node.children:
                    queue.append(n)
            res.append(row_res)
        return res

        
# @lc code=end

